File: architecture/modules/storage/node/src/service.ts
import { randomUUID } from 'node:crypto';
import { StorageError,validSize,validContentType,ttl } from './contracts.ts';
import type {PutInput,ObjectDownload} from './contracts.ts';
import type {FileInfo,UploadInput,UploadSession} from './client.ts';
import type {FileRecord,FileRepository} from './repository.ts';
import {StorageRouter} from './router.ts';
import {readable,sizeGuard} from './streams.ts';
import {pipeline} from 'node:stream/promises';
import {PassThrough} from 'node:stream';

const visible=(r:FileRecord):FileInfo=>({id:r.id,name:r.name,size:r.size,contentType:r.contentType,state:r.state,createdAt:r.createdAt,updatedAt:r.updatedAt});
export class FileService {
  readonly router:StorageRouter;readonly repository:FileRepository;readonly maxSize:number;readonly lifetime:number;readonly clock:()=>number;
  constructor(options:{router:StorageRouter;repository:FileRepository;maxSize?:number;expiresIn?:number;clock?:()=>number}){
    this.router=options.router;this.repository=options.repository;this.maxSize=validSize(options.maxSize??32*1024*1024);this.lifetime=ttl(options.expiresIn??900);this.clock=options.clock||Date.now;
  }
  private async own(ownerId:string,id:string):Promise<FileRecord>{
    if(!ownerId||ownerId.length>255)throw new StorageError('FORBIDDEN');
    const r=await this.repository.get(id);if(!r||r.ownerId!==ownerId)throw new StorageError('NOT_FOUND');return r;
  }
  private async move(record:FileRecord,fields:Partial<FileRecord>):Promise<FileRecord>{
    const next={...record,...fields,version:record.version+1,updatedAt:new Date(this.clock()).toISOString()};
    if(!await this.repository.compareAndSwap(record.id,record.version,next))throw new StorageError('CONFLICT','File changed; read its state before retry');return next;
  }
  private active(r:FileRecord){if(Date.parse(r.expiresAt)<=this.clock())throw new StorageError('EXPIRED');}
  async createUpload(ownerId:string,input:UploadInput):Promise<UploadSession>{
    if(!ownerId||ownerId.length>255)throw new StorageError('FORBIDDEN');
    if(!input||typeof input.name!=='string'||!input.name.trim()||input.name.length>255||/[\x00-\x1f\x7f]/.test(input.name))throw new StorageError('INVALID_INPUT','Invalid file name');
    validSize(input.size,this.maxSize);validContentType(input.contentType);const storeId=input.storeId||this.router.defaultStore,provider=this.router.get(storeId);
    const id=randomUUID(),now=new Date(this.clock()).toISOString(),expiresAt=new Date(this.clock()+this.lifetime*1000).toISOString();
    const record:FileRecord={schemaVersion:1,id,ownerId,storeId,objectKey:'files/'+id+'/'+randomUUID(),tempKey:'uploads/'+id+'/'+randomUUID(),name:input.name,size:input.size,contentType:input.contentType,state:'pending',version:1,createdAt:now,updatedAt:now,expiresAt};
    const signed=provider.signUpload?await provider.signUpload(record.tempKey,{size:record.size,contentType:record.contentType,expiresIn:this.lifetime}):undefined;
    await this.repository.create(record);
    return {file:visible(record),expiresAt,transport:signed?{...signed,method:'PUT',kind:'signed'}:{kind:'proxy',method:'PUT',path:'/uploads/'+id+'/content'}};
  }
  async upload(ownerId:string,id:string,body:PutInput['body'],signal?:AbortSignal):Promise<void>{
    let r=await this.own(ownerId,id);this.active(r);if(r.state!=='pending')throw new StorageError('CONFLICT');
    r=await this.move(r,{state:'uploading',leaseUntil:this.clock()+300000});
    const operation=AbortSignal.any([AbortSignal.timeout(300000),...(signal?[signal]:[])]);
    try{await this.router.get(r.storeId).put(r.tempKey,{body,size:r.size,contentType:r.contentType,signal:operation});await this.move(r,{state:'pending',leaseUntil:undefined});}catch(e){await this.move(r,{state:'pending',leaseUntil:undefined}).catch(()=>{});throw e;}
  }
  async complete(ownerId:string,id:string,signal?:AbortSignal):Promise<FileInfo>{
    let r=await this.own(ownerId,id);if(r.state==='ready')return visible(r);this.active(r);if(r.state!=='pending')throw new StorageError('CONFLICT');
    r=await this.move(r,{state:'completing',leaseUntil:this.clock()+300000});const provider=this.router.get(r.storeId),operation=AbortSignal.any([AbortSignal.timeout(300000),...(signal?[signal]:[])]);
    try{
      const download=await provider.get(r.tempKey,{signal:operation});if(download.info.size!==r.size||download.info.contentType!==r.contentType){download.body.destroy();throw new StorageError('INVALID_INPUT','Uploaded metadata differs from session');}
      const bridge=new PassThrough(),guard=sizeGuard(r.size);
      const sending=provider.put(r.objectKey,{body:bridge,size:r.size,contentType:r.contentType,signal:operation});
      const copying=pipeline(download.body,guard.stream,bridge,{signal:operation});
      try{await Promise.all([copying,sending]);}catch(e){bridge.destroy();download.body.destroy();await Promise.allSettled([copying,sending]);throw e;}
      r=await this.move(r,{state:'ready',leaseUntil:undefined});
      // Temporary URL can still be replayed; only the separate final object is downloadable.
      await provider.delete(r.tempKey).catch(()=>{});
      return visible(r);
    }catch(e){
      // Leave completing for explicit recovery; an uncertain DB write must never delete a ready object.
      throw e;
    }
  }
  async recover(ownerId:string,id:string):Promise<FileInfo>{
    let r=await this.own(ownerId,id);
    if(['completing','uploading'].includes(r.state)){
      if((r.leaseUntil||0)>this.clock())throw new StorageError('CONFLICT','Operation lease is still active');
      const previousState=r.state;
      r=await this.move(r,{leaseUntil:this.clock()+300000});
      if(previousState==='completing')await this.router.get(r.storeId).delete(r.objectKey);
      r=await this.move(r,{state:'pending',objectKey:'files/'+r.id+'/'+randomUUID(),leaseUntil:undefined});
    }else if(r.state==='deleting'){await this.removeObjects(r);r=await this.move(r,{state:'deleted'});}
    return visible(r);
  }
  async cancel(ownerId:string,id:string):Promise<void>{
    let r=await this.own(ownerId,id);if(r.state==='cancelled'){await this.router.get(r.storeId).delete(r.tempKey);return;}if(r.state!=='pending')throw new StorageError('CONFLICT');
    r=await this.move(r,{state:'cancelled'});await this.router.get(r.storeId).delete(r.tempKey);
  }
  async info(ownerId:string,id:string):Promise<FileInfo>{return visible(await this.own(ownerId,id));}
  async download(ownerId:string,id:string,signal?:AbortSignal):Promise<ObjectDownload>{const r=await this.own(ownerId,id);if(r.state!=='ready')throw new StorageError('NOT_FOUND');return this.router.get(r.storeId).get(r.objectKey,{signal});}
  async downloadLocation(ownerId:string,id:string):Promise<{url?:string;headers?:Record<string,string>;expiresAt?:string;path?:string}>{
    const r=await this.own(ownerId,id);if(r.state!=='ready')throw new StorageError('NOT_FOUND');const p=this.router.get(r.storeId);
    if(p.signDownload){const s=await p.signDownload(r.objectKey,{expiresIn:Math.min(this.lifetime,300)});return {url:s.url,headers:s.headers,expiresAt:s.expiresAt};}return {path:'/'+id+'/content'};
  }
  private async removeObjects(r:FileRecord){const p=this.router.get(r.storeId);await p.delete(r.objectKey);await p.delete(r.tempKey);}
  async delete(ownerId:string,id:string):Promise<void>{
    let r=await this.own(ownerId,id);if(r.state==='deleted')return;if(r.state==='ready')r=await this.move(r,{state:'deleting'});else if(r.state!=='deleting')throw new StorageError('CONFLICT');
    await this.removeObjects(r);await this.move(r,{state:'deleted'});
  }
  async list(ownerId:string,input:{cursor?:string;limit?:number}={}){if(!ownerId)throw new StorageError('FORBIDDEN');const page=await this.repository.list(ownerId,input);return {...page,items:page.items.map(visible)};}
}

File: architecture/modules/storage/go/service.go
package storage

import (
	"context"
	"io"
	"strings"
	"time"
)

type UploadInput struct {
	Name        string `json:"name"`
	Size        int64  `json:"size"`
	ContentType string `json:"contentType"`
	StoreID     string `json:"storeId,omitempty"`
}
type UploadTransport struct {
	Kind      string            `json:"kind"`
	Method    string            `json:"method"`
	Path      string            `json:"path,omitempty"`
	URL       string            `json:"url,omitempty"`
	Headers   map[string]string `json:"headers,omitempty"`
	ExpiresAt string            `json:"expiresAt,omitempty"`
}
type UploadSession struct {
	File      FileInfo        `json:"file"`
	ExpiresAt string          `json:"expiresAt"`
	Transport UploadTransport `json:"transport"`
}
type FileService struct {
	Router     *Router
	Repository Repository
	MaxSize    int64
	Lifetime   time.Duration
	Now        func() time.Time
}

func NewFileService(router *Router, repository Repository) *FileService {
	return &FileService{Router: router, Repository: repository, MaxSize: 32 << 20, Lifetime: 15 * time.Minute, Now: time.Now}
}
func (s *FileService) own(ctx context.Context, owner, id string) (FileRecord, error) {
	if owner == "" || len(owner) > 255 {
		return FileRecord{}, fail("FORBIDDEN")
	}
	r, e := s.Repository.Get(ctx, id)
	if e != nil {
		return r, e
	}
	if r.OwnerID != owner {
		return FileRecord{}, fail("NOT_FOUND")
	}
	return r, nil
}
func (s *FileService) move(ctx context.Context, r FileRecord, update func(*FileRecord)) (FileRecord, error) {
	next := r
	update(&next)
	next.Version++
	next.UpdatedAt = s.Now().UTC().Format(time.RFC3339Nano)
	ok, e := s.Repository.CompareAndSwap(ctx, r.ID, r.Version, next)
	if e != nil {
		return r, e
	}
	if !ok {
		return r, fail("CONFLICT")
	}
	return next, nil
}
func (s *FileService) active(r FileRecord) error {
	expiry, e := time.Parse(time.RFC3339Nano, r.ExpiresAt)
	if e != nil || !expiry.After(s.Now()) {
		return fail("EXPIRED")
	}
	return nil
}
func (s *FileService) CreateUpload(ctx context.Context, owner string, input UploadInput) (UploadSession, error) {
	if owner == "" || len(owner) > 255 {
		return UploadSession{}, fail("FORBIDDEN")
	}
	if strings.TrimSpace(input.Name) == "" || len(input.Name) > 255 || strings.ContainsAny(input.Name, "\r\n\x00") || input.Size > s.MaxSize {
		return UploadSession{}, fail("INVALID_INPUT")
	}
	if e := validInput(PutInput{Body: strings.NewReader(""), Size: input.Size, ContentType: input.ContentType}); e != nil {
		return UploadSession{}, e
	}
	if _, e := duration(s.Lifetime, Credentials{}); e != nil {
		return UploadSession{}, e
	}
	store := input.StoreID
	if store == "" {
		store = s.Router.DefaultStore
	}
	p, e := s.Router.Get(store)
	if e != nil {
		return UploadSession{}, e
	}
	id := uuid()
	now := s.Now().UTC()
	r := FileRecord{FileInfo: FileInfo{ID: id, Name: input.Name, Size: input.Size, ContentType: input.ContentType, State: "pending", CreatedAt: now.Format(time.RFC3339Nano), UpdatedAt: now.Format(time.RFC3339Nano)}, SchemaVersion: 1, OwnerID: owner, StoreID: store, ObjectKey: "files/" + id + "/" + uuid(), TempKey: "uploads/" + id + "/" + uuid(), ExpiresAt: now.Add(s.Lifetime).Format(time.RFC3339Nano), Version: 1}
	t := UploadTransport{Kind: "proxy", Method: "PUT", Path: "/uploads/" + id + "/content"}
	if p.Capabilities().DirectUpload {
		signed, e := p.SignUpload(ctx, r.TempKey, r.Size, r.ContentType, s.Lifetime)
		if e != nil {
			return UploadSession{}, e
		}
		t = UploadTransport{Kind: "signed", Method: "PUT", URL: signed.URL, Headers: signed.Headers, ExpiresAt: signed.ExpiresAt}
	}
	if e = s.Repository.Create(ctx, r); e != nil {
		return UploadSession{}, e
	}
	return UploadSession{File: r.FileInfo, ExpiresAt: r.ExpiresAt, Transport: t}, nil
}
func (s *FileService) Upload(ctx context.Context, owner, id string, body io.Reader) error {
	r, e := s.own(ctx, owner, id)
	if e != nil {
		return e
	}
	if e = s.active(r); e != nil {
		return e
	}
	if r.State != "pending" {
		return fail("CONFLICT")
	}
	r, e = s.move(ctx, r, func(n *FileRecord) { n.State = "uploading"; n.LeaseUntil = s.Now().Add(5 * time.Minute).UnixMilli() })
	if e != nil {
		return e
	}
	p, e := s.Router.Get(r.StoreID)
	if e != nil {
		return e
	}
	operation, cancel := context.WithTimeout(ctx, 5*time.Minute)
	defer cancel()
	_, e = p.Put(operation, r.TempKey, PutInput{Body: body, Size: r.Size, ContentType: r.ContentType})
	if e != nil {
		return e
	}
	_, e = s.move(ctx, r, func(n *FileRecord) { n.State = "pending"; n.LeaseUntil = 0 })
	return e
}
func (s *FileService) Complete(ctx context.Context, owner, id string) (FileInfo, error) {
	r, e := s.own(ctx, owner, id)
	if e != nil {
		return FileInfo{}, e
	}
	if r.State == "ready" {
		return r.FileInfo, nil
	}
	if e = s.active(r); e != nil {
		return FileInfo{}, e
	}
	if r.State != "pending" {
		return FileInfo{}, fail("CONFLICT")
	}
	r, e = s.move(ctx, r, func(n *FileRecord) { n.State = "completing"; n.LeaseUntil = s.Now().Add(5 * time.Minute).UnixMilli() })
	if e != nil {
		return FileInfo{}, e
	}
	p, e := s.Router.Get(r.StoreID)
	if e != nil {
		return FileInfo{}, e
	}
	operation, cancel := context.WithTimeout(ctx, 5*time.Minute)
	defer cancel()
	d, e := p.Get(operation, r.TempKey)
	if e != nil {
		return FileInfo{}, e
	}
	defer d.Body.Close()
	if d.Info.Size != r.Size || d.Info.ContentType != r.ContentType {
		return FileInfo{}, fail("INVALID_INPUT")
	}
	_, e = p.Put(operation, r.ObjectKey, PutInput{Body: d.Body, Size: r.Size, ContentType: r.ContentType})
	if e != nil {
		return FileInfo{}, e
	}
	r, e = s.move(ctx, r, func(n *FileRecord) { n.State = "ready"; n.LeaseUntil = 0 })
	if e != nil {
		return FileInfo{}, e
	}
	_ = p.Delete(ctx, r.TempKey)
	return r.FileInfo, nil
}
func (s *FileService) Recover(ctx context.Context, owner, id string) (FileInfo, error) {
	r, e := s.own(ctx, owner, id)
	if e != nil {
		return FileInfo{}, e
	}
	if r.State == "uploading" || r.State == "completing" {
		if r.LeaseUntil > s.Now().UnixMilli() {
			return FileInfo{}, fail("CONFLICT")
		}
		state := r.State
		r, e = s.move(ctx, r, func(n *FileRecord) { n.LeaseUntil = s.Now().Add(5 * time.Minute).UnixMilli() })
		if e != nil {
			return FileInfo{}, e
		}
		if state == "completing" {
			p, e := s.Router.Get(r.StoreID)
			if e != nil {
				return FileInfo{}, e
			}
			if e = p.Delete(ctx, r.ObjectKey); e != nil {
				return FileInfo{}, e
			}
		}
		r, e = s.move(ctx, r, func(n *FileRecord) {
			n.State = "pending"
			n.ObjectKey = "files/" + n.ID + "/" + uuid()
			n.LeaseUntil = 0
		})
		if e != nil {
			return FileInfo{}, e
		}
	} else if r.State == "deleting" {
		if e = s.Delete(ctx, owner, id); e != nil {
			return FileInfo{}, e
		}
		return s.Info(ctx, owner, id)
	}
	return r.FileInfo, nil
}
func (s *FileService) Cancel(ctx context.Context, owner, id string) error {
	r, e := s.own(ctx, owner, id)
	if e != nil {
		return e
	}
	if r.State == "pending" {
		r, e = s.move(ctx, r, func(n *FileRecord) { n.State = "cancelled" })
		if e != nil {
			return e
		}
	} else if r.State != "cancelled" {
		return fail("CONFLICT")
	}
	p, e := s.Router.Get(r.StoreID)
	if e != nil {
		return e
	}
	return p.Delete(ctx, r.TempKey)
}
func (s *FileService) Info(ctx context.Context, owner, id string) (FileInfo, error) {
	r, e := s.own(ctx, owner, id)
	return r.FileInfo, e
}
func (s *FileService) Download(ctx context.Context, owner, id string) (Download, error) {
	r, e := s.own(ctx, owner, id)
	if e != nil {
		return Download{}, e
	}
	if r.State != "ready" {
		return Download{}, fail("NOT_FOUND")
	}
	p, e := s.Router.Get(r.StoreID)
	if e != nil {
		return Download{}, e
	}
	return p.Get(ctx, r.ObjectKey)
}
func (s *FileService) DownloadLocation(ctx context.Context, owner, id string) (map[string]any, error) {
	r, e := s.own(ctx, owner, id)
	if e != nil {
		return nil, e
	}
	if r.State != "ready" {
		return nil, fail("NOT_FOUND")
	}
	p, e := s.Router.Get(r.StoreID)
	if e != nil {
		return nil, e
	}
	if p.Capabilities().DirectUpload {
		signed, e := p.SignDownload(ctx, r.ObjectKey, 5*time.Minute)
		if e != nil {
			return nil, e
		}
		return map[string]any{"url": signed.URL, "headers": signed.Headers, "expiresAt": signed.ExpiresAt}, nil
	}
	return map[string]any{"path": "/" + id + "/content"}, nil
}
func (s *FileService) Delete(ctx context.Context, owner, id string) error {
	r, e := s.own(ctx, owner, id)
	if e != nil {
		return e
	}
	if r.State == "deleted" {
		return nil
	}
	if r.State == "ready" {
		r, e = s.move(ctx, r, func(n *FileRecord) { n.State = "deleting" })
		if e != nil {
			return e
		}
	} else if r.State != "deleting" {
		return fail("CONFLICT")
	}
	p, e := s.Router.Get(r.StoreID)
	if e != nil {
		return e
	}
	if e = p.Delete(ctx, r.ObjectKey); e != nil {
		return e
	}
	if e = p.Delete(ctx, r.TempKey); e != nil {
		return e
	}
	_, e = s.move(ctx, r, func(n *FileRecord) { n.State = "deleted" })
	return e
}
func (s *FileService) List(ctx context.Context, owner, cursor string, n int) (FilePage, error) {
	if owner == "" {
		return FilePage{}, fail("FORBIDDEN")
	}
	page, e := s.Repository.List(ctx, owner, cursor, n)
	if e != nil {
		return FilePage{}, e
	}
	result := FilePage{Items: []FileInfo{}, Cursor: page.Cursor}
	for _, r := range page.Items {
		result.Items = append(result.Items, r.FileInfo)
	}
	return result, nil
}
